File: algorithms/medium/maximum_of_absolute_value_expression.cpp
class Solution {
public:
  int maxAbsValExpr(vector<int> &arr1, vector<int> &arr2) {
    vector<array<int, 4>> mat(arr1.size());
    for (int i = 0; i < arr1.size(); i++) {
      array<int, 4> arr = {arr1[i] + arr2[i] + i, arr1[i] + arr2[i] - i,
                           arr1[i] - arr2[i] + i, arr1[i] - arr2[i] - i};
      mat[i] = move(arr);
    }

    int ans = 0;
    for (int j = 0; j < 4; j++) {
      int ma = INT_MIN, mi = INT_MAX;
      for (int i = 0; i < mat.size(); i++) {
        ma = max(mat[i][j], ma);
        mi = min(mat[i][j], mi);
      }
      ans = max(ans, ma - mi);
    }

    return ans;
  }
};

File: algorithms/medium/maximum_of_absolute_value_expression.ex
defmodule Solution do
  @spec max_abs_val_expr(arr1 :: [integer], arr2 :: [integer]) :: integer
  def max_abs_val_expr(arr1, arr2) do
    Enum.zip(arr1, arr2)
    |> Enum.with_index()
    |> Enum.reduce([], fn {{x, y}, i}, acc ->
      [[x + y + i, x - y + i, x + y - i, x - y - i] | acc]
    end)
    |> Enum.zip()
    |> Enum.max_by(&diff/1)
    |> then(&diff/1)
  end

  def diff(tp) do
    {mi, ma} = tp |> Tuple.to_list() |> Enum.min_max()
    ma - mi
  end
end
